python/graphstorm/model/embed.py

```diff
@@ -290,29 +354,46 @@ def forward(self, input_feats, input_nodes):
         assert isinstance(input_nodes, dict), 'The input node IDs should be in a dict.'
         embs = {}
         for ntype in input_nodes:
+            if isinstance(input_nodes[ntype], np.ndarray):
+                # WholeGraphSparseEmbedding requires the input nodes (indexing tensor)
+                # to be a th.Tensor
+                input_nodes[ntype] = th.from_numpy(input_nodes[ntype])
             emb = None
             if ntype in input_feats:
                 assert ntype in self.input_projs, \
-                        f"We need a projection for node type {ntype}"
+                    f"We need a projection for node type {ntype}"
                 # If the input data is not float, we need to convert it t float first.
                 emb = input_feats[ntype].float() @ self.input_projs[ntype]
                 if self.use_node_embeddings:
                     assert ntype in self.sparse_embeds, \
-                            f"We need sparse embedding for node type {ntype}"
-                    node_emb = self.sparse_embeds[ntype](input_nodes[ntype], emb.device)
-                    concat_emb=th.cat((emb, node_emb),dim=1)
+                        f"We need sparse embedding for node type {ntype}"
+                    # emb.device: target device to put the gathered results
+                    if self._use_wholegraph_sparse_emb:
+                        node_emb = self.sparse_embeds[ntype].module(input_nodes[ntype].cuda())
```

Comment:
Right, the output tensor will be on the current cuda device. 

>> Shall we use to(emb.device) instead of cuda()? Can cuda move the tensor to the right GPU?
As long as we called torch.cuda.set_device, `cuda()` indicates the current cuda device. In fact, this is a constraint of WholeGraph that it requires the input indexes can only located at the current cuda device (here, `module()` function is very similar to wholegraph `gather()` call).

Reply:
Right, it's already in GPU. But I am not sure if `emb.device` must be current cuda device, or could it be `cpu`?

Reply:
I c. it is in GPU.